src/pinns_pde_solve.jl

```diff
@@ -0,0 +1,250 @@
+
+    # coefficients for loss function
+    τb = length(train_bound_set)
+    τf = length(train_domain_set)
+
+    #dimensionality of equation
+    dim = prob.dim
+
+    dim > 3 && error("While only dimensionality no more than 3")
+
+    # neural network
+    chain  = alg.chain
+    # optimizer
+    opt    = alg.opt
+    # AD flag
+    autodiff = alg.autodiff
+    # weights of NN
+    initθ = alg.initθ
+
+    # for a system of equations, not yet supported
+    isuinplace = true
+
+    # The phi trial solution
+    if chain isa FastChain
+        if isuinplace
+            phi = (x,θ) -> first(chain(adapt(typeof(θ),x),θ))
+        else
+            phi = (x,θ) -> chain(adapt(typeof(θ),x),θ)
+        end
+    else
+        _,re  = Flux.destructure(chain)
+        if isuinplace
+            phi = (x,θ) -> first(re(θ)(adapt(typeof(θ),x)))
+        else
+            phi = (x,θ) -> re(θ)(adapt(typeof(θ),x))
+        end
+    end
+
+    # calculate derivative
+    if autodiff #automatic differentiation (not implemented yet)
+        # derivative = (x,θ,n) -> ForwardDiff.gradient(x->phi(x,θ),x)[n]
+    else # numerical derivative
+        ep = cbrt(eps(Float64))
+        eps_masks = [[[ep]],
+                    [[ep,0.0], [0.0,ep]],
+                    [[ep,0.0,0.0], [0.0,ep,0.0],[0.0,0.0,ep]]]
+
+        derivative = (_x...) ->
+        begin
+            u_in = _x[1]
+            x = collect(_x[2:end-2])
+            der_num =_x[end-1]
+            θ = _x[end]
+            (phi(x+eps_masks[dim][der_num], θ) - phi(x,θ))/ep
+        end
+        second_order_derivative = (_x...) ->
+        begin
+            u_in = _x[1]
+            x = collect(_x[2:end-2])
+            der_num =_x[end-1]
+            θ = _x[end]
+            (phi(x+eps_masks[dim][der_num], θ) - 2*phi(x,θ)+ phi(x-eps_masks[dim][der_num], θ) )/(ep^2)
+        end
+    end
+
+    #represent pde function
+    _u= (_x...; x = collect(_x[1:end-1]),θ = _x[end]) -> phi(x,θ)
+    pde_func = (_x,θ) -> _pde_func(_u,_x...,θ,derivative,second_order_derivative)
+
+
+    #loss function for pde equation
+    function inner_loss_domain(x,θ)
+        pde_func(x,θ)
+    end
+
+    function loss_domain(θ)
+        sum(abs2,inner_loss_domain(x,θ) for x in train_domain_set)
+    end
+
+    #loss function for Dirichlet boundary
+    function inner_loss(x,θ,bound)
+        phi(x,θ) - bound
+    end
+
+    # #Neumann boundary
+    # function inner_neumann_loss(x,θ,num,bound)
+    #     derivative(u,x,num θ) - bound
+    # end
+
+    #loss function for boundary condiiton
+    function loss_boundary(θ)
+       sum(abs2,inner_loss(x,θ,bound) for (x,bound) in train_bound_set)
+    end
+
+    #loss function
+    loss(θ) = 1.0f0/τf * loss_domain(θ) + 1.0f0/τb * loss_boundary(θ) #+ 1.0f0/τc * custom_loss(θ)
+
+    cb = function (p,l)
+        verbose && println("Current loss is: $l")
+        l < abstol
+    end
+    res = DiffEqFlux.sciml_train(loss, initθ, opt; cb = cb, maxiters=maxiters, alg.kwargs...)
```

Comment:
I am registering GalacticOptim https://github.com/JuliaRegistries/General/pull/17124 which is supposed to replace the `sciml_train` interface, you would want to make a note to switch to that

Reply:
ok, I will keep it in mind